### PR TITLE
reconciler: don't delete join token based entries

### LINKED ISSUE #247: Entry reconciler deletes entries corresponding to join tokens
The entry reconciler deletes registration entries corresponding to join tokens. Due to this, any join tokens created on the server managed by the controller is unusable. Should there be an exception to entries corresponding to join tokens in the entry reconciler? We could potentially avoid cleanup of join token entries by ensuring entries whose parent SPIFFE ID contains `/spire/agent/join_token/` aren't reconciled.

### PR DESCRIPTION
**Pull request by shashankram**
Updates the reconciler to ignore the deletion of join token based entries since join tokens are created out-of-band and can be used in conjunction with the controller-manager.

Resolves #247

### CODE REVIEW

File: pkg/spireentry/reconciler.go [hunk 1]
```diff
@@ -630,3 +640,31 @@ func idsFromEntries(entries []spireapi.Entry) []string {
...
+	if !strings.Contains(entry.ParentID.String(), joinTokenSpiffeSubstr) {
+		return false
+	}
+	for _, selector := range entry.Selectors {
+		if selector.Type == joinTokenSelectorType {
+			return true
+		}
+	}
+	return false
+}
```
Comment (kfox1111) [inline pkg/spireentry/reconciler.go:670]: Is this code used?

Same with the definition joinTokenSelectorType

**Review by kfox1111 — COMMENTED** [review]
Looks good to me. One issue inline.

File: pkg/spireentry/reconciler.go (hunk 1 above)
Comment (shashankram) [reply pkg/spireentry/reconciler.go:670]: good catch, forgot to use the helper. Fixed now.

**Comment by shashankram** [comment]
> Looks good to me. One issue inline.

fixed, thanks

File: pkg/spireentry/reconciler.go [hunk 2]
```diff
@@ -630,3 +640,31 @@ func idsFromEntries(entries []spireapi.Entry) []string {
 	}
 	return ids
 }
+
+// filterJoinTokenEntries filters out entries that correspond to join tokens.
+func filterJoinTokenEntries(entries []spireapi.Entry) []spireapi.Entry {
+	var filteredEntries []spireapi.Entry
```
Comment (azdagron) [inline pkg/spireentry/reconciler.go:646]: nit: we should take the opportunity to pre-allocate the slice to reduce allocation pressure
```suggestion
	if len(entries) == 0 {
		return entries
	}	
	filteredEntries := make([]spireapi.Entry, 0, len(entries))
```

File: pkg/spireentry/reconciler.go [hunk 3]
```diff
@@ -630,3 +640,31 @@ func idsFromEntries(entries []spireapi.Entry) []string {
...
+	}
+	return filteredEntries
+}
+
+// isJoinTokenEntry returns true if the entry corresponds to a join token.
+// For an entry to correspond to a join token, both the following conditions must be true:
+// 1. The parent ID of the entry must contain the substring "/spire/agent/join_token/".
+// 2. The entry must contain a selector of type "spiffe_id".
+func isJoinTokenEntry(entry spireapi.Entry) bool {
+	if !strings.Contains(entry.ParentID.String(), joinTokenSpiffeSubstr) {
```
Comment (azdagron) [inline pkg/spireentry/reconciler.go:661]:
```suggestion
	if !strings.HasPrefix(entry.ParentID.Path(), joinTokenSpiffeSubstr) {
```

File: pkg/spireentry/reconciler.go [hunk 4]
```diff
@@ -630,3 +640,31 @@ func idsFromEntries(entries []spireapi.Entry) []string {
...
+			continue
+		}
+		filteredEntries = append(filteredEntries, entry)
+	}
+	return filteredEntries
+}
+
+// isJoinTokenEntry returns true if the entry corresponds to a join token.
+// For an entry to correspond to a join token, both the following conditions must be true:
+// 1. The parent ID of the entry must contain the substring "/spire/agent/join_token/".
```
Comment (azdagron) [inline pkg/spireentry/reconciler.go:658]:
```suggestion
// 1. The path of the parent ID of the entry must begin with "/spire/agent/join_token/".
```

**Review by azdagron — COMMENTED** [review]
Thanks @shashankram. Just a few small comments.

**Review by azdagron — DISMISSED** [review]

**Review by azdagron — APPROVED** [review]

**Pull request merged**